soroban-env-common/src/raw_val.rs

```diff
@@ -1,101 +1,163 @@
-use stellar_xdr::{ScStatic, ScStatus, ScStatusType};
+use crate::{impl_wrapper_tryfroms_and_tryfromvals, SymbolSmall};
+use stellar_xdr::{ScStatus, ScStatusType};
 
-use super::{BitSet, Env, Object, Static, Status, Symbol, TryFromVal};
+use super::{Env, Object, Status, TryFromVal};
 use core::{convert::Infallible, fmt::Debug};
 
 extern crate static_assertions as sa;
 
+// A Soroban Val is a 64-bit word with the low 8 bits assigned to a `tag`
+// indicating its type and the high 56 bits reserved for its `body`
+
 #[allow(dead_code)]
 const WORD_BITS: usize = 64;
-const TAG_BITS: usize = 3;
+pub(crate) const TAG_BITS: usize = 8;
 const TAG_MASK: u64 = (1u64 << TAG_BITS) - 1;
-sa::const_assert!(TAG_MASK == 0x7);
+sa::const_assert!(TAG_MASK == 0xff);
 
 #[allow(dead_code)]
-pub(crate) const BODY_BITS: usize = WORD_BITS - (TAG_BITS + 1);
-sa::const_assert!(BODY_BITS == 60);
+pub(crate) const BODY_BITS: usize = WORD_BITS - TAG_BITS;
+sa::const_assert!(BODY_BITS == 56);
+
+// The body is sometimes further subdivided into two fields:
+// a 32-bit `major` part and a 24-bit `minor` part.
 
-#[allow(dead_code)]
 const MAJOR_BITS: usize = 32;
-#[allow(dead_code)]
-const MINOR_BITS: usize = 28;
-#[allow(dead_code)]
+const MINOR_BITS: usize = 24;
 const MAJOR_MASK: u64 = (1u64 << MAJOR_BITS) - 1;
 const MINOR_MASK: u64 = (1u64 << MINOR_BITS) - 1;
 sa::const_assert!(MAJOR_MASK == 0xffff_ffff);
-sa::const_assert!(MINOR_MASK == 0x0fff_ffff);
+sa::const_assert!(MINOR_MASK == 0x00ff_ffff);
 sa::const_assert!(MAJOR_BITS + MINOR_BITS == BODY_BITS);
 
-/// Code values for the 3 "tag" bits in the bit-packed representation
-/// of [RawVal].
+/// Code values for the 8 `tag` bits in the bit-packed representation
```

Comment:
In the `stellar-contract.x` it was mentioned the maximum value types is 128 (7 bits), is it because the 8-bit tag value space here is a superset of the xdr value types? Or do we only intend to use 7 of the 8 bit tag bits here?

Reply:
Only 7 in practice -- actually I think we're going to come in under 6 -- but I went with 1 byte since there's almost always some part of a computer system that prefers 8 bit to 7 bit quantities.